src/treetensornetworks/projttns/abstractprojttn.jl

```diff
@@ -131,13 +131,43 @@ end
 function position(
   P::AbstractProjTTN{V}, psi::TTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
 ) where {V}
-  # shift position
   P = shift_position(P, pos)
-  # remove internal edges (out of place)
+  P = invalidate_environments(P)
+  P = make_environments(P, psi)
+  return P
+end
+
+function position!(
+  P::AbstractProjTTN{V}, psi::TTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
+) where {V}
+  P = shift_position(P, pos)
+  P = invalidate_environments(P)
+  make_environments!(P, psi, e)
+  return P
+end
+
+function invalidate_environments(P::AbstractProjTTN)
   ie = internal_edges(P)
   newenvskeys = filter(!in(ie), keys(P.environments))
```

Comment:
I think we should define a function `environments(p::AbstractProjTTN)` which accesses `p.environments`, generally I try to define those kinds of field accessor functions instead of accessing fields with the `.` syntax, which I think looks cleaner and helps with writing generic code.

Reply:
Similarly, I think it would be good to define functions `operator(p::ProjTTN) = p.H` and `position(p::ProjTTN) = p.pos` and use those in the code.

Reply:
Still seems to be using `P.environments` instead of `environments(P)`.